### DISCUSSION TITLE
Native file/directory picker

### DISCUSSION THREAD

**Discussion by tbillington** [emilk/egui#987] [Q&A] [answered]
Hi! I'm wondering how I'd allow the user to open a file/folder through egui, whether that's through launching the native file picker widget or some other mechanism. I searched the docs on egui, eframe, and glow but none of my searches turned up anything.

**Comment by coderedart** [comment, marked as answer]
You use https://github.com/PolyMeilex/rfd .

NOTE: there's some concerns like this being blocking the whole UI (should not be a concern usually, and when it is, threads/channels should be helpful) and also something with MacOS (a simple search in issues section of egui and rfd should help you find more about this and other things).

**Reply by emilk** [reply]
There is an example of using `rfd` with `eframe` at https://github.com/emilk/egui/blob/master/eframe/examples/file_dialog.rs

**Reply by yuribudilov** [reply]
FYI, FWIW, I am getting error "status heap corruption" error on Windows-10 (0xc0000374) when I drag a (rfd::) file to the egui window. I am using egui/eframe/epi (0.16.0) and rfd (0.7.0), latest as of Feb 2022. I am not really able to use any of the provided eframe examples beyond the first most basic one. The example generally do not compile (missing functions or classes or both) and when I do change the code they dont run correctly. Please correct me if I am wrong, I am totally new to egui/eframe/epi and the documentation does not have enough help/examples for the new-comers to egui/eframe/epi to proceed further. I am able to make really basic apps work OK but nothing further. Thanks.

**Reply by yuribudilov** [reply]
More: (debugging further) the Windows-10 heap corruption happens if I click, drag file name to the egui window. If I left-click, Select on menu, it works. So the way I select/click/drag file names is what causes the heap corruption error. Perhaps it is related to the examples source code fragment handling of the dropped_files vector. Perhaps I am doing it all wrong but with the incomplete code examples it is very hard to proceed given I am totally new to egui/eframe/epi, trying to learn it from scratch.

**Reply by emilk** [reply]
Heap corruption is a very uncommon error in rust, and should only happen due to `unsafe` code, which `egui` has none of.

I suggest firing up a debugger (and run a memcheck on your RAM).

**Reply by yuribudilov** [reply]
thank you very much.
